### PR TITLE
feat(sermant-springboot-registry): 添加 Nacos 注册中心支持

### PR DESCRIPTION
**Pull request by xzhaoxz**
What type of PR is this?

spring-boot supports nacos

What this PR does / why we need it?

SpringBoot provides service registration and discovery capabilities for pure SpringBoot applications, enabling users to quickly access the registry without modifying the code

Which issue(s) this PR fixes？

Fixes #

Does this PR introduce a user-facing change?

#1714

### CODE REVIEW

File: sermant-plugins/sermant-springboot-registry/config/config.yaml [hunk 1]
```diff
@@ -7,6 +7,8 @@ sermant.springboot.registry:
   realmName: www.domain.com
   # Whether to enable traffic statistics, after it is enabled, the traffic entering the plug-in will be printed every time
   enableRequestCount: false
+  # The type of the registry center, currently supports Nacos and Zookeeper
+  registryCenterType:
```
Comment (daizhenyu) [inline sermant-plugins/sermant-springboot-registry/config/config.yaml]: Use all uppercase, NACOS or ZOOKEEPER, and registryCenterType should be given default value

File: sermant-plugins/sermant-springboot-registry/config/config.yaml [hunk 2]
```diff
@@ -28,3 +30,19 @@ sermant.springboot.registry.lb:
   # For retry scenarios, for {@link java.util.concurrent.TimeoutException}, whether you need to retry, which is enabled
   # by default, and this timeout is mostly used in asynchronous scenarios, such as Future, MinimalHttpAsyncClient
   enableTimeoutExRetry: true
+
+nacos.service:
+  # nacos verifies accounts
+  username: "nacos"
+  # nacos verifies password need encrypts
+  password: "nacos"
```
Comment (daizhenyu) [inline sermant-plugins/sermant-springboot-registry/config/config.yaml]: remove password value, and note could be nacos encrypted password.

File: sermant-plugins/sermant-springboot-registry/springboot-registry-plugin/src/main/java/io/sermant/discovery/config/DiscoveryPluginConfig.java [hunk 3]
```diff
@@ -42,6 +42,11 @@ public class DiscoveryPluginConfig implements PluginConfig {
      */
     private boolean enableRegistry = true;
 
+    /**
+     * Registry type, currently only ZK is supported
```
Comment (daizhenyu) [inline sermant-plugins/sermant-springboot-registry/springboot-registry-plugin/src/main/java/io/sermant/discovery/config/DiscoveryPluginConfig.java]: wrong comment

File: sermant-plugins/sermant-springboot-registry/config/config.yaml [hunk 4]
```diff
@@ -28,3 +30,19 @@ sermant.springboot.registry.lb:
...
+  # Specify the AES privateKey that encrypts and decrypts the password
+  privateKey:
+  # Namespace, the nacos configuration creates the id value of the namespace
+  namespace: "sermant"
+  # Service instance weight value
+  weight: 1
+  # Cluster name
+  clusterName: DEFAULT
+  # Whether it is a temporary node. The value can be true or false
+  ephemeral: true
```
Comment (daizhenyu) [inline sermant-plugins/sermant-springboot-registry/config/config.yaml:50]: Is the service a temporary node by default?

File: sermant-plugins/sermant-springboot-registry/springboot-registry-plugin/src/main/java/io/sermant/discovery/config/NacosRegisterConfig.java [hunk 5]
```diff
@@ -0,0 +1,400 @@
...
+
+    /**
+     * privateKey
+     */
+    private String privateKey;
+
+    /**
+     * Node address
+     */
+    private String endpoint = "";
```
Comment (daizhenyu) [inline sermant-plugins/sermant-springboot-registry/springboot-registry-plugin/src/main/java/io/sermant/discovery/config/NacosRegisterConfig.java]: Is endpoint the address of the host microservice?

File: sermant-plugins/sermant-springboot-registry/springboot-registry-plugin/src/main/java/io/sermant/discovery/config/NacosRegisterConfig.java [hunk 6]
```diff
@@ -0,0 +1,400 @@
...
+
+    /**
+     * Whether it fails to retrieve cached data quickly, if false is not retrieved, it will fail directly
+     */
+    private boolean failureToleranceEnabled = false;
+
+    /**
+     * Service name separator
+     */
+    private String serviceNameSeparator = ":";
```
Comment (daizhenyu) [inline sermant-plugins/sermant-springboot-registry/springboot-registry-plugin/src/main/java/io/sermant/discovery/config/NacosRegisterConfig.java]: put it into constant class

File: sermant-plugins/sermant-springboot-registry/springboot-registry-plugin/src/main/java/io/sermant/discovery/config/NacosRegisterConfig.java [hunk 7]
```diff
@@ -0,0 +1,400 @@
...
+
+    /**
+     * Get configuration parameters
+     *
+     * @return Properties
+     */
+    public Properties getNacosProperties() {
+        LbConfig lbConfig = PluginConfigManager.getPluginConfig(LbConfig.class);
+        Properties properties = new Properties();
+        if (StringUtils.isEmpty(username) || StringUtils.isEmpty(password)
```
Comment (daizhenyu) [inline sermant-plugins/sermant-springboot-registry/springboot-registry-plugin/src/main/java/io/sermant/discovery/config/NacosRegisterConfig.java]: Both authorization and non-authorization modes should be provided. you can add a switch of authorization

File: sermant-plugins/sermant-springboot-registry/springboot-registry-plugin/src/main/java/io/sermant/discovery/config/NacosRegisterConfig.java [hunk 8]
```diff
@@ -0,0 +1,400 @@
...
+        if (!pwd.isPresent()) {
+            LOGGER.log(Level.SEVERE, "Nacos password parsing failed");
+            return properties;
+        }
+        properties.put(PropertyKeyConst.SERVER_ADDR, lbConfig.getRegistryAddress());
+        properties.put(PropertyKeyConst.USERNAME, Objects.toString(username, ""));
+        properties.put(PropertyKeyConst.PASSWORD, pwd.get());
+        properties.put(PropertyKeyConst.NAMESPACE, Objects.toString(namespace, ""));
+        properties.put(PropertyKeyConst.NACOS_NAMING_LOG_NAME, Objects.toString(logName, ""));
+        if (endpoint.contains(PropertyKeyConst.HTTP_URL_COLON)) {
```
Comment (daizhenyu) [inline sermant-plugins/sermant-springboot-registry/springboot-registry-plugin/src/main/java/io/sermant/discovery/config/NacosRegisterConfig.java]: why endpoint have two format

File: sermant-plugins/sermant-springboot-registry/springboot-registry-plugin/src/main/java/io/sermant/discovery/entity/RegisterContext.java [hunk 9]
```diff
@@ -40,4 +42,122 @@ public DefaultServiceInstance getServiceInstance() {
...
 
+    private final ClientInfo clientInfo = new ClientInfo();
+
+    public ClientInfo getClientInfo() {
+        return clientInfo;
+    }
+
+
+    /**
+     * 客户端信息
```
Comment (daizhenyu) [inline sermant-plugins/sermant-springboot-registry/springboot-registry-plugin/src/main/java/io/sermant/discovery/entity/RegisterContext.java]: use english comment

File: sermant-plugins/sermant-springboot-registry/springboot-registry-service/src/main/java/io/sermant/discovery/service/lb/discovery/nacos/NacosDiscoveryClient.java [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Locale;
+import java.util.Optional;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * nacos Register plug-in configuration
```
Comment (daizhenyu) [inline sermant-plugins/sermant-springboot-registry/springboot-registry-service/src/main/java/io/sermant/discovery/service/lb/discovery/nacos/NacosDiscoveryClient.java]: update comment

File: sermant-plugins/sermant-springboot-registry/springboot-registry-service/src/main/java/io/sermant/discovery/service/lb/discovery/zk/ZkClient.java [hunk 11]
```diff
@@ -51,7 +51,11 @@ public ZkClient() {
 
     @Override
     public void start() {
-        if (!PluginConfigManager.getPluginConfig(DiscoveryPluginConfig.class).isEnableRegistry()) {
+        final String registryCenterType = PluginConfigManager.getPluginConfig(DiscoveryPluginConfig.class)
```
Comment (daizhenyu) [inline sermant-plugins/sermant-springboot-registry/springboot-registry-service/src/main/java/io/sermant/discovery/service/lb/discovery/zk/ZkClient.java]: registryCenterType could use enum

File: sermant-plugins/sermant-springboot-registry/springboot-registry-service/src/main/java/io/sermant/discovery/service/lb/discovery/zk/ZkClient.java [hunk 12]
```diff
@@ -51,7 +51,11 @@ public ZkClient() {
...
+                .getRegistryCenterType();
+        String zkName = "Zookeeper";
+        if (!PluginConfigManager.getPluginConfig(DiscoveryPluginConfig.class).isEnableRegistry()
```
Comment (daizhenyu) [inline sermant-plugins/sermant-springboot-registry/springboot-registry-service/src/main/java/io/sermant/discovery/service/lb/discovery/zk/ZkClient.java]: the result of PluginConfigManager.getPluginConfig(DiscoveryPluginConfig.class)  should save by temporary variable，only get once

File: sermant-plugins/sermant-springboot-registry/springboot-registry-service/src/main/java/io/sermant/discovery/service/lb/rule/NacosWeightRandomRule.java [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+
+    /**
+     * 根据权重随机
+     * 算法参考 https://blog.csdn.net/u011627980/article/details/79401026
+     *
+     * @param list 实例列表
+     * @return 随机出来的结果
+     */
+    public ServiceInstance weightRandom(List<InstanceWithWeight> list) {
+        List<ServiceInstance> instances = Lists.newArrayList();
```
Comment (daizhenyu) [inline sermant-plugins/sermant-springboot-registry/springboot-registry-service/src/main/java/io/sermant/discovery/service/lb/rule/NacosWeightRandomRule.java]: please optimize this algorithm ，refer：
    int totalWeight = 0;
    for (InstanceWithWeight instanceWithWeight : list) {
        totalWeight += instanceWithWeight.getWeight();
    }
    int randomWeight = new Random().nextInt(totalWeight);
    int currentWeight = 0;
    for (InstanceWithWeight instanceWithWeight : list) {
        currentWeight += instanceWithWeight.getWeight();
        if (randomWeight < currentWeight) {
            return instanceWithWeight.getServer();
        }
    }

File: sermant-plugins/sermant-springboot-registry/springboot-registry-service/src/test/java/io/sermant/discovery/service/lb/discovery/nacos/NacosDiscoveryClientTest.java [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+import io.sermant.discovery.entity.RegisterContext;
+import io.sermant.discovery.entity.ServiceInstance;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.*;
+import org.mockito.junit.MockitoJUnitRunner;
+
+import java.util.*;
```
Comment (daizhenyu) [inline sermant-plugins/sermant-springboot-registry/springboot-registry-service/src/test/java/io/sermant/discovery/service/lb/discovery/nacos/NacosDiscoveryClientTest.java]: not import *, import a specific class.

File: sermant-plugins/sermant-springboot-registry/springboot-registry-service/src/test/java/io/sermant/discovery/service/lb/discovery/nacos/NacosDiscoveryClientTest.java [hunk 15]
```diff
@@ -0,0 +1,117 @@
...
+    @Mock
+    private NamingService namingService;
+
+    private NacosRegisterConfig registerConfig = new NacosRegisterConfig();
+
+    private NacosDiscoveryClient nacosDiscoveryClient;
+
+    private NacosServiceManager nacosServiceManager;
+
+    private MockedStatic<PluginConfigManager> pluginConfigManagerMockedStatic;
```
Comment (daizhenyu) [inline sermant-plugins/sermant-springboot-registry/springboot-registry-service/src/test/java/io/sermant/discovery/service/lb/discovery/nacos/NacosDiscoveryClientTest.java:42]: Check whether close mock is required in teardown method.

File: sermant-plugins/sermant-springboot-registry/springboot-registry-plugin/src/main/java/io/sermant/discovery/config/NacosRegisterConfig.java [hunk 16]
```diff
@@ -0,0 +1,402 @@
...
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * nacos Register plug-in configuration
+ *
+ * @author xiaozhao
+ * @since 2024-11-16
+ */
+@ConfigTypeKey(value = "nacos.service")
```
Comment (lilai23) [inline sermant-plugins/sermant-springboot-registry/springboot-registry-plugin/src/main/java/io/sermant/discovery/config/NacosRegisterConfig.java]: The value of @ConfigTypeKey can not be the same as other plugins.  I read the code and find that currently When use ConfigManager.getConfig(NacosRegisterConfig .class); It may get wrong config

File: sermant-plugins/sermant-springboot-registry/springboot-registry-plugin/src/main/java/io/sermant/discovery/config/NacosRegisterConfig.java [hunk 17]
```diff
@@ -0,0 +1,402 @@
...
+    /**
+     * Default pull interval
+     */
+    private static final long DEFAULT_NOTIFY_DELAY = 5000L;
+
+    /**
+     * The default monitoring time
+     */
+    private static final long DEFAULT_LOOKUP_INTERVAL = 30L;
+
```
Comment (lilai23) [inline sermant-plugins/sermant-springboot-registry/springboot-registry-plugin/src/main/java/io/sermant/discovery/config/NacosRegisterConfig.java]: Specify time unit

File: sermant-plugins/sermant-springboot-registry/springboot-registry-plugin/src/main/java/io/sermant/discovery/config/NacosRegisterConfig.java [hunk 18]
```diff
@@ -0,0 +1,402 @@
...
+            properties.put(PropertyKeyConst.ENDPOINT, endpoint.substring(0, index));
+            properties.put(PropertyKeyConst.ENDPOINT_PORT, endpoint.substring(index + 1));
+        } else {
+            properties.put(PropertyKeyConst.ENDPOINT, endpoint);
+        }
+        properties.put(PropertyKeyConst.ACCESS_KEY, Objects.toString(accessKey, ""));
+        properties.put(PropertyKeyConst.SECRET_KEY, Objects.toString(secretKey, ""));
+        properties.put(PropertyKeyConst.CLUSTER_NAME, clusterName);
+        properties.put(PropertyKeyConst.NAMING_LOAD_CACHE_AT_START, namingLoadCacheAtStart);
+        return properties;
```
Comment (lilai23) [inline sermant-plugins/sermant-springboot-registry/springboot-registry-plugin/src/main/java/io/sermant/discovery/config/NacosRegisterConfig.java]: endpoint is never set and default value is "", check is it necessary to config this.  And same as other config keys

File: sermant-plugins/sermant-springboot-registry/springboot-registry-service/src/test/java/io/sermant/discovery/service/lb/rule/NacosWeightRandomRuleTest.java [hunk 19]
```diff
@@ -0,0 +1,66 @@
...
+
+import io.sermant.discovery.entity.ServiceInstance;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.InjectMocks;
+import org.junit.Assert;
+import org.mockito.junit.MockitoJUnitRunner;
+
+import java.util.*;
```
Comment (daizhenyu) [inline sermant-plugins/sermant-springboot-registry/springboot-registry-service/src/test/java/io/sermant/discovery/service/lb/rule/NacosWeightRandomRuleTest.java]: import splitly

File: sermant-plugins/sermant-springboot-registry/springboot-registry-service/src/test/java/io/sermant/discovery/service/lb/rule/NacosWeightRandomRuleTest.java [hunk 20]
```diff
@@ -0,0 +1,66 @@
...
+    }
+
+    @Test
+    public void lbType() {
+        Assert.assertEquals(new NacosWeightRandomRule().lbType(), "NacosWeight");
+    }
+
+    @Test
+    public void testWeightRandom() {
+        List<NacosWeightRandomRule.InstanceWithWeight> withWeights = instances.stream()
```
Comment (daizhenyu) [inline sermant-plugins/sermant-springboot-registry/springboot-registry-service/src/test/java/io/sermant/discovery/service/lb/rule/NacosWeightRandomRuleTest.java:60]: Write more test cases to deal with boundary conditions

File: sermant-plugins/sermant-springboot-registry/springboot-registry-service/src/main/java/io/sermant/discovery/service/lb/rule/NacosWeightRandomRule.java [hunk 21]
```diff
@@ -0,0 +1,95 @@
...
+        int totalWeight = list.stream().mapToInt(InstanceWithWeight::getWeight).sum();
+        int randomWeight = new Random().nextInt(totalWeight);
+        int currentWeight = 0;
+        for (InstanceWithWeight instanceWithWeight : list) {
+            currentWeight += instanceWithWeight.getWeight();
+            if (randomWeight < currentWeight) {
+                return instanceWithWeight.getServer();
+            }
+        }
+        return list.get(0).getServer();
```
Comment (daizhenyu) [inline sermant-plugins/sermant-springboot-registry/springboot-registry-service/src/main/java/io/sermant/discovery/service/lb/rule/NacosWeightRandomRule.java]: This is actually unreachable. Should we throw an exception or print a log?

File: sermant-plugins/sermant-springboot-registry/springboot-registry-service/src/main/java/io/sermant/discovery/service/lb/rule/NacosWeightRandomRule.java [hunk 22]
```diff
@@ -0,0 +1,95 @@
...
+ *
+ * @author xiaozhao
+ * @since 2024-11-16
+ */
+public class NacosWeightRandomRule extends AbstractLoadbalancer {
+    @Override
+    protected ServiceInstance doChoose(String serviceName, List<ServiceInstance> instances) {
+        List<InstanceWithWeight> withWeights = instances.stream()
+                .map(instance -> {
+                    int weight = Integer.parseInt(instance.getMetadata().get("nacos.weight"));
```
Comment (zwmagic) [inline sermant-plugins/sermant-springboot-registry/springboot-registry-service/src/main/java/io/sermant/discovery/service/lb/rule/NacosWeightRandomRule.java]: instance.getMetadata().get("nacos.weight") may be null?

File: sermant-plugins/sermant-springboot-registry/springboot-registry-service/src/main/java/io/sermant/discovery/service/lb/discovery/nacos/listen/NacosInstanceListenable.java [hunk 23]
```diff
@@ -0,0 +1,114 @@
...
+            return;
+        }
+        List<ServiceInstance> serviceInstances = new ArrayList<>(instances.size());
+        for (Instance instance : instances) {
+            if (instance.isEnabled() && instance.isHealthy()) {
+                Optional<ServiceInstance> optional = nacosServiceManager.convertServiceInstance(instance, serviceName);
+                optional.ifPresent(serviceInstances::add);
+            }
+        }
+        listenerCache.get(serviceName).notify(serviceInstances);
```
Comment (zwmagic) [inline sermant-plugins/sermant-springboot-registry/springboot-registry-service/src/main/java/io/sermant/discovery/service/lb/discovery/nacos/listen/NacosInstanceListenable.java]: listenerCache.get(serviceName) may be null?

File: sermant-plugins/sermant-springboot-registry/springboot-registry-service/src/main/java/io/sermant/discovery/service/lb/discovery/nacos/NacosDiscoveryClient.java [hunk 24]
```diff
@@ -0,0 +1,161 @@
...
+
+    @Override
+    public void init() {
+    }
+
+    @Override
+    public boolean registry(ServiceInstance serviceInstance) {
+        String serviceId = serviceInstance.getServiceName();
+        String group = nacosRegisterConfig.getGroup();
+        instance = nacosServiceManager.buildNacosInstanceFromRegistration(serviceInstance);
```
Comment (zwmagic) [inline sermant-plugins/sermant-springboot-registry/springboot-registry-service/src/main/java/io/sermant/discovery/service/lb/discovery/nacos/NacosDiscoveryClient.java:73]: Should we consider the case where the host application has already been using the Nacos client?

**Review by daizhenyu — APPROVED** [review]

**Review by lilai23 — APPROVED** [review]
